controller/scheduler-v2/scheduler.go

```diff
@@ -0,0 +1,1190 @@
+	// all current hosts before returning) *before* registering in
+	// service discovery so that there is always at least one scheduler
+	// watching all job events, even during a deployment.
+	if err := s.streamHostEvents(); err != nil {
+		return err
+	}
+
+	var err error
+	s.isLeader, err = s.discoverd.Register()
+	if err != nil {
+		return err
+	}
+	leaderCh := s.discoverd.LeaderCh()
+
+	if err := s.streamFormationEvents(); err != nil {
+		return err
+	}
+
+	s.tickSyncJobs(30 * time.Second)
+	s.tickSyncFormations(time.Minute)
+
+	go s.RunPutJobs()
+
+	for {
+		select {
+		case <-s.stop:
+			log.Info("stopping scheduler loop")
+			close(s.putJobs)
+			return nil
+		case isLeader := <-leaderCh:
+			s.HandleLeaderChange(isLeader)
+			continue
+		default:
+		}
+
+		// Handle events that reconcile scheduler state with the cluster
+		select {
+		case req := <-s.jobRequests:
+			s.HandleJobRequest(req)
+			continue
+		case e := <-s.hostEvents:
+			s.HandleHostEvent(e)
+			continue
+		case e := <-s.jobEvents:
+			s.HandleJobEvent(e)
+			continue
+		case f := <-s.formationEvents:
+			s.HandleFormationChange(f)
+			continue
+		default:
+		}
+
+		// Handle sync events
+		select {
+		case <-s.syncFormations:
+			s.SyncFormations()
+			continue
+		case <-s.syncJobs:
+			s.SyncJobs()
+			continue
+		default:
+		}
+
+		// Finally, handle triggering cluster changes.
+		// Re-select on all the channels so we don't have to sleep nor spin
+		select {
+		case <-s.rectify:
+			s.Rectify()
+		case <-s.stop:
+			log.Info("stopping scheduler loop")
+			close(s.putJobs)
+			return nil
+		case isLeader := <-leaderCh:
+			s.HandleLeaderChange(isLeader)
+		case req := <-s.jobRequests:
+			s.HandleJobRequest(req)
+		case e := <-s.hostEvents:
+			s.HandleHostEvent(e)
+		case e := <-s.jobEvents:
+			s.HandleJobEvent(e)
+		case f := <-s.formationEvents:
+			s.HandleFormationChange(f)
+		case <-s.syncFormations:
+			s.SyncFormations()
+		case <-s.syncJobs:
+			s.SyncJobs()
+		}
+	}
+	return nil
+}
+
+func (s *Scheduler) SyncJobs() {
+	defer s.sendEvent(NewEvent(EventTypeClusterSync, nil, nil))
+
+	log := fnLogger()
+	log.Info("syncing jobs")
+
+	log.Info("getting host list")
+	hosts, err := s.getHosts()
+	if err != nil {
+		log.Error("error getting host list", "err", err)
+		return
+	}
+
+	newJobs := make(Jobs)
+	for _, h := range hosts {
+		hostLog := log.New("host.id", h.ID())
+
+		hostLog.Info(fmt.Sprintf("getting jobs for host %s", h.ID()))
+		activeJobs, err := h.ListJobs()
+		if err != nil {
+			hostLog.Error("error getting jobs list", "err", err)
+			continue
+		}
+		hostLog.Info(fmt.Sprintf("got %d active job(s) for host %s", len(activeJobs), h.ID()))
+
+		for _, job := range activeJobs {
+			s.handleActiveJob(&job)
+			if j, ok := s.jobs[job.Job.ID]; ok {
+				newJobs[j.JobID] = s.jobs[j.JobID]
+			}
+		}
+	}
+
+	for id, j := range s.jobs {
+		if _, ok := newJobs[id]; !ok && j.IsRunning() {
+			s.handleJobStop(j)
+		}
```

Comment:
What is this trying to catch? It seems as though it is marking all jobs in `s.jobs` as stopped which no longer appear as an active job from any host in the cluster, which will only happen if a host which was previously running jobs no longer appears in `cluster.Hosts`. Is this the intention?